hmiddleware/connection_closer.go

```diff
@@ -0,0 +1,75 @@
+package hmiddleware
+
+import (
+	"context"
+	"log"
+	"net"
+	"net/http"
+	"sync"
+
+	"github.com/google/uuid"
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/heroku/x/go-kit/metrics"
+)
+
+type connCloseKey int
+
+var connCloseID connCloseKey
+
+// ConnectionClosingContext adds a unique identifier to the context for a newly established connection. This function
+// is meant to be used as an http.Server's ConnContext field, and is required for the ConnectionClosingMiddleware.
+func ConnectionClosingContext(ctx context.Context, c net.Conn) context.Context {
+	return context.WithValue(ctx, connCloseID, uuid.New().String())
+}
+
+// ConnectionClosingMiddleware returns middleware that will add a "Connection: close" responses header after a single
+// connection has produced maxRequests http requests. The cacheSize determines the size of the LRU cache which tracks
+// connection request counts. This middleware requires the http.Server sets ConnectionClosingContext for it's
+// ConnContext field.
+func ConnectionClosingMiddleware(mp metrics.Provider, maxRequests, cacheSize int) func(http.Handler) http.Handler {
+	var (
+		closed = mp.NewCounter("server.connection.closes.total")
+		mtx    sync.Mutex
+	)
+
+	cache, err := lru.New(cacheSize)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			shouldClose := false
+			connID, ok := r.Context().Value(connCloseID).(string)
+
+			if ok {
+				// In the case of http/1.x requests are served serially per connection, and locking isn't necessary.
+				// However this isn't necessarily true for the http/2 path. So we lock here for correctness.
+				mtx.Lock()
+
+				requests := 0
+				requestValue, ok := cache.Get(connID)
+				if ok {
+					requests = requestValue.(int)
+				}
+
+				requests++
+				if requests >= maxRequests {
+					shouldClose = true
+					cache.Remove(connID)
+				} else {
+					cache.Add(connID, requests)
+				}
+
+				mtx.Unlock()
+			}
+
+			next.ServeHTTP(w, r)
+
+			if shouldClose {
+				closed.Add(1)
+				w.Header().Add("connection", "close")
```

Comment:
If I understand correctly, this can only be called here if next didn't write anything to w. Should this be done before next.ServeHTTP is called on what will be the last request?

Reply:
I guess I may not be 100% clear on the conventions for this connection close header. Is it expected that there'd be no other content to read from a response in the presence of the header? e.g. we should flip things to do this (i think this is what you're suggesting):

```
if shouldClose {
    closed.Add(1)
    w.Header().Add("connection", "close")
    return // <-- add this
}

// they're not at the limit yet, call next handler
next.ServeHTTP(w, r)

```

Reply:
If you can say "this will be the last request" I think it's fine to set the header and then call next.ServeHTTP. But if you call next.ServeHTTP first and then try and set the header I'm not sure it will work if w.WriteHeader has been called (or any body has been written). Maybe the http.Server bits will still see `Connection: close` and close the connection, I'm not sure. Using httptest.NewServer instead of a ResponseRecorder to get a more complete test rig might help build confidence.

Reply:
after poking at the net/http code this seems to be done cooperatively. whether the header is present in an http request from a client, or if its set in a response header, the desire to close the connection is respected in net/http. the idea here is once we hit the threshold we return the response like usual, but add an additional `connection: close` header.